plugins/rest/rest_auth.go

```diff
@@ -122,6 +104,112 @@ func (ap *bearerAuthPlugin) Prepare(req *http.Request) error {
 	return nil
 }
 
+type tokenEndpointResponse struct {
+	AccessToken string `json:"access_token"`
+	TokenType   string `json:"token_type"`
+	ExpiresIn   int64  `json:"expires_in"`
+}
+
+// oauth2ClientCredentialsAuthPlugin represents authentication via a bearer token in the HTTP Authorization header
+// obtained through the OAuth2 client credentials flow
+type oauth2ClientCredentialsAuthPlugin struct {
+	TokenURL     string    `json:"token_url"`
+	ClientID     string    `json:"client_id"`
+	ClientSecret string    `json:"client_secret"`
+	Scopes       *[]string `json:"scopes,omitempty"`
+
+	tokenCache    *oauth2Token
+	tlsSkipVerify bool
+}
+
+type oauth2Token struct {
+	Token     string
+	ExpiresAt time.Time
+}
+
+func (ap *oauth2ClientCredentialsAuthPlugin) NewClient(c Config) (*http.Client, error) {
+	t, err := defaultTLSConfig(c)
+	if err != nil {
+		return nil, err
+	}
+
+	// Inherit skip verify from the "parent" settings. Should this be configurable on the credentials too?
+	ap.tlsSkipVerify = c.AllowInsureTLS
+
+	if !strings.HasPrefix(ap.TokenURL, "https://") {
+		return nil, errors.New("token_url required to use https scheme")
+	}
```

Comment:
Was there a specific reason it needed to be https? Playing around with keycloak and a token url of `http://localhost:8080/auth/realms/myrealm/protocol/openid-connect/token` seemed to work fine after I removed this check.

Reply:
Just blindly following [spec](https://tools.ietf.org/html/rfc6749#section-10.3) here :) 

> Access token credentials (as well as any confidential access token attributes) MUST be kept confidential in transit and storage, and only shared among the authorization server, the resource servers the access token is valid for, and the client to whom the access token is issued. Access token credentials MUST only be transmitted using TLS as described in Section 1.6 with server authentication as defined by [RFC2818].

Arguably though this is more the responsibility of the server that the client, so I'm fine with dropping this requirement. I think we should at least log a warning here though if the token endpoint is configured without https.

Reply:
Ahh ok. If it is part of the spec then I'm cool with leaving the check in.